tensorflow_probability/python/experimental/mcmc/streaming_reductions_test.py

```diff
@@ -0,0 +1,161 @@
+        counter_1=previous_kernel_results.counter_1 + 1,
+        counter_2=previous_kernel_results.counter_2 + 2)
+
+  def bootstrap_results(self, current_state):
+    return TestTransitionKernelResults(counter_1=0, counter_2=0)
+
+  @property
+  def is_calibrated(self):
+    return self._is_calibrated
+
+
+class RandomTransitionKernel(tfp.mcmc.TransitionKernel):
+
+  def __init__(self, is_calibrated=True, accepts_seed=True):
+    self._is_calibrated = is_calibrated
+    self._accepts_seed = accepts_seed
+
+  def one_step(self, current_state, previous_kernel_results, seed=None):
+    if seed is not None and not self._accepts_seed:
+      raise TypeError('seed arg not accepted')
+    random_next_state = tfp.random.rayleigh((1,), seed=seed)
+    return random_next_state, previous_kernel_results
+
+  @property
+  def is_calibrated(self):
+    return self._is_calibrated
+
+
+@test_util.test_all_tf_execution_regimes
+class StepKernelTest(test_util.TestCase):
+
+  def test_simple_operation(self):
+    fake_kernel = TestTransitionKernel()
+    final_state, kernel_results = step_kernel(
+        num_steps=2,
+        current_state=0,
+        kernel=fake_kernel,
+        return_final_kernel_results=True
+    )
+    final_state = self.evaluate(final_state)
+    self.assertEqual(final_state, 2)
+    self.assertEqual(kernel_results.counter_1, 2)
+    self.assertEqual(kernel_results.counter_2, 4)
+
+  def test_defined_pkr(self):
+    fake_kernel = TestTransitionKernel()
+    final_state, kernel_results = step_kernel(
+        num_steps=2,
+        current_state=0,
+        previous_kernel_results=TestTransitionKernelResults(2, 3),
+        kernel=fake_kernel,
+        return_final_kernel_results=True
+    )
+    final_state = self.evaluate(final_state)
+    self.assertEqual(final_state, 2)
+    self.assertEqual(kernel_results.counter_1, 4)
+    self.assertEqual(kernel_results.counter_2, 7)
+
+  @parameterized.parameters(1, 2)
+  def test_initial_states(self, init_state):
+    fake_kernel = TestTransitionKernel()
+    final_state = step_kernel(
+        num_steps=2,
+        current_state=init_state,
+        kernel=fake_kernel,
+    )
+    final_state = self.evaluate(final_state)
+    self.assertEqual(final_state, init_state + 2)
+
+  def test_calibration_warning(self):
+    with warnings.catch_warnings(record=True) as triggered:
+      kernel = TestTransitionKernel(is_calibrated=False)
+      tfp.mcmc.sample_chain(
+          num_results=2,
+          current_state=0,
+          kernel=kernel,
+          trace_fn=lambda current_state, kernel_results: kernel_results,
+          seed=test_util.test_seed())
+    self.assertTrue(
+        any('supplied `TransitionKernel` is not calibrated.' in str(
+            warning.message) for warning in triggered))
+
+  def test_seed_reproducibility(self):
+    first_fake_kernel = RandomTransitionKernel()
+    second_fake_kernel = RandomTransitionKernel()
+    seed = test_util.test_seed()
+    if tf.executing_eagerly():
```

Comment:
This if statement is necessary so the test also runs correctly in graph mode

Reply:
I expect that sanitizing unconditionally should work in both.  Do you know what goes wrong?

Note that one of the weird features of TF Classic PRNG semantics is the effect under multiple calls to `self.evaluate`.  The safe thing to do is to only ever compare results from the same `self.evaluate` call.  In this test, that would look like
```
last_state_t = step_kernel...  # No evaluate here
for num_steps in range
  first_final_state_t = step_kernel...  # Not here either
  second_final_state_t = step_kernel...  # Nor here
  last_state, first_final_state, second_final_state = self.evaluate([last_state_t, first_final_state_t, second_final_state_t])
  self.assertEqual ...
```
I'm actually not sure what the `for` loop will do here.  Ought to work as I suggested, but we could also fall back to building up a big list of Tensors, evaluating them once outside the `for`, and then writing another `for` for the assertions.

Reply:
that did the trick!